crates/cli/src/main.rs

```diff
@@ -1,52 +1,226 @@
-use anyhow::{anyhow, bail, Result};
+use anyhow::{anyhow, bail, Context, Ok, Result};
 use clap::Parser;
 use core::fmt::Write;
-use std::fs;
+use std::{
+    ffi::OsStr,
+    fs,
+    path::{Path, PathBuf},
+};
 use wasmi::{
     core::{Value, ValueType, F32, F64},
+    Engine,
     ExportItemKind,
     Func,
     FuncType,
+    Linker,
+    Module,
     Store,
 };
+use wasmi_wasi::{ambient_authority, Dir, TcpListener, WasiCtx, WasiCtxBuilder};
+
+fn parse_env(s: &str) -> Result<(String, String)> {
+    s.split_once('=')
+        .map(|(s1, s2)| (s1.to_owned(), s2.to_owned()))
+        .ok_or_else(|| anyhow::anyhow!("expects '='-separated pair"))
+}
 
 /// Simple program to greet a person
 #[derive(Parser, Debug)]
 #[clap(author, version, about, long_about = None)]
 struct Args {
-    /// The WebAssembly file to execute.
-    #[clap(value_parser)]
-    wasm_file: String,
+    /// The host directory to pre-open for the `guest` to use.
+    #[clap(long = "dir", number_of_values = 1)]
+    dirs: Vec<String>,
+
+    /// The socket address provided to the module. Allows it to perform socket-related `WASI` ops.
+    #[clap(long = "tcplisten", number_of_values = 1)]
+    tcplisten: Vec<String>,
+
+    /// The environment variable pair made available for the program.
+    #[clap(long = "env", value_parser(parse_env))]
+    envs: Vec<(String, String)>,
```

Comment:
```suggestion
    #[clap(value_parser = parse_env, action = clap::ArgAction::Append)]
    env: Option<KeyValue>,
```

Reply:
This requires us to add a `KeyValue` struct which is cleaner than `(String, String)`:
```rust
pub struct KeyValue {
    key: String,
    value: String,
}
```

Reply:
@Robbepop , I noticed that you changed `env:  Vec<KeyValue>` into `env: Option<KeyValue>` in your reviews. Did you remember that we need `env`, `tcplisten`, and `dir` to contain multiple values?

Reply:
Yes, I do but to mimic Wasmtime CLI behavior properly we use `action = clap::ArgAction::Append` as commented above. I also provided some links to the `clap` manual describing usage of `Append` option.